src/System.IO.Compression/src/System/IO/Compression/InflaterZlib.cs

```diff
@@ -0,0 +1,275 @@
+
+            lock (_syncLock)
+            {
+                _inputBufferHandle = GCHandle.Alloc(inputBuffer, GCHandleType.Pinned);
+                _zlibStream.NextIn = _inputBufferHandle.Value.AddrOfPinnedObject() + startIndex;
+                _zlibStream.AvailIn = (uint)count;
+                _finished = false;
+            }
+        }
+
+        [SecuritySafeCritical]
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_isDisposed && disposing)
+            {
+                if (_inputBufferHandle.HasValue)
+                    DeallocateInputBufferHandle();
+
+                _zlibStream.Dispose();
+                _isDisposed = true;
+            }
+        }
+
+        void IDisposable.Dispose()
+        {
+            Dispose(true);
+        }
+
+        #endregion
+
+        #region Helper Methods
+
+        /// <summary>
+        /// Creates the ZStream that will handle inflation
+        /// </summary>
+        [SecuritySafeCritical]
+        private void InflateInit(int windowBits)
+        {
+            ZLibNative.ErrorCode error;
+            try
+            {
+                error = ZLibNative.CreateZLibStreamForInflate(out _zlibStream, windowBits);
+            }
+            catch (Exception exception) // could not load the ZLib dll
+            {
+                throw new ZLibException(SR.ZLibErrorDLLLoadError, exception);
+            }
+
+            switch (error)
+            {
+                case ZLibNative.ErrorCode.Ok:           // Successful initialization
+                    return;
+
+                case ZLibNative.ErrorCode.MemError:     // Not enough memory
+                    throw new ZLibException(SR.ZLibErrorNotEnoughMemory, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                case ZLibNative.ErrorCode.VersionError: //zlib library is incompatible with the version assumed
+                    throw new ZLibException(SR.ZLibErrorVersionMismatch, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                case ZLibNative.ErrorCode.StreamError:  // Parameters are invalid
+                    throw new ZLibException(SR.ZLibErrorIncorrectInitParameters, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                default:
+                    throw new ZLibException(SR.ZLibErrorUnexpected, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+            }
+        }
+
+        /// <summary>
+        /// Translates the given byte array to a GCHandle so that it can be passed to the ZLib
+        /// Inflate function, then returns the result of that call.
+        /// </summary>
+        private ZLibNative.ErrorCode ReadInflateOutput(byte[] outputBuffer, int offset, int length, ZLibNative.FlushCode flushCode, out int bytesRead)
+        {
+            lock (_syncLock)
+            {
+                GCHandle outputBufferHndl = GCHandle.Alloc(outputBuffer, GCHandleType.Pinned);
```

Comment:
> is this fine and if so, which context would be preferred to be marked as such?

That's fine.  I think it'd be fine just to add unsafe to the method signature, since effectively the whole method needs to be able to use pointers.

> does fixed preclude the necessity of the lock, or do they accomplish unrelated things?

They are orthogonal, can be used together, and are both needed.  The lock ensures that only one thread at a time can access this region.  The fixed ensures that for the duration of the fixed block, the buffer is pinned, and it gives you a pointer to it.

> so I'll be mirroring changes from your feedback to that file as well.

Sounds good. Thanks.